File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/user/UserService.java
package top.likoslupus.cellulosesz.api.user;

import org.jspecify.annotations.Nullable;
import top.likoslupus.cellulosesz.api.platform.CellPlayer;

import java.util.Collection;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;
import java.util.function.UnaryOperator;

public interface UserService {

    CompletableFuture<CellUser> load(UUID uuid);

    CompletableFuture<CellUser> loadFromPlayer(CellPlayer player);

    Optional<CellUser> cached(UUID uuid);

    Collection<CellUser> cachedUsers();

    Optional<UUID> findUuidByName(String name);

    Collection<UUID> knownUuids();

    default CompletableFuture<Void> updateVoid(
            UUID uuid,
            UnaryOperator<CellUser> mutation
    ) {
        return update(
                uuid,
                user -> UserUpdate.replacing(mutation.apply(user))
        ).thenAccept(_ -> {
        });
    }

    <T extends @Nullable Object> CompletableFuture<T> update(
            UUID uuid,
            Function<CellUser, UserUpdate<T>> mutation
    );

    CompletableFuture<Void> save(UUID uuid);

    CompletableFuture<Void> saveAll();

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/user/UserService.kt
package top.likoslupus.cellulosesz.api.user

import top.likoslupus.cellulosesz.api.platform.CellPlayer
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.function.Function
import java.util.function.UnaryOperator

public interface UserService {

    public fun load(uuid: UUID): CompletableFuture<CellUser>

    public fun loadFromPlayer(player: CellPlayer): CompletableFuture<CellUser>

    public fun cached(uuid: UUID): CellUser?

    public fun cachedUsers(): Collection<CellUser>

    public fun findUuidByName(name: String): UUID?

    public fun knownUuids(): Collection<UUID>

    public fun updateVoid(
        uuid: UUID,
        mutation: UnaryOperator<CellUser>
    ): CompletableFuture<Void> =
        update(uuid) { user ->
            UserUpdate.replacing(mutation.apply(user))
        }.thenAccept { }

    public fun <T> update(
        uuid: UUID,
        mutation: Function<CellUser, UserUpdate<T>>
    ): CompletableFuture<T>

    public fun save(uuid: UUID): CompletableFuture<Void>

    public fun saveAll(): CompletableFuture<Void>

}
